design.md

```diff
@@ -0,0 +1,106 @@
+* `rpc GetOuput(GetOutputRequest) returns (stream JobOutput) {}` - Returns the process output stream given the job id. A `NOT_FOUND` error is returned if the provided job id does not exist. 
+    * If the job is running, `GetOutput(...)` replays all previous output if any, from the job before streaming new messages. The stream is closed when the job reaches a final state
+    * If called when the process is in a final state, `GetOutput(...)` sends all job output and closes the stream.
+    
+## Client/CLI
+The cli is the interface for the user to send commands to the server. The user must provide the client certificate, client key, CA certificate and server address to run the command. Assuming the name of the cli executable is teleport, then the command format is: 
+```shell
+> teleport --cacert="/path/to/ca.cert" \
+           --key="/path/to/client.key" \ 
+           --cert="/path/to/client.cert" \
+           --addr="server address" 
+           COMMAND [OPTIONS] [ARG...]
+ ```
+`[Options]` is the set of flags or options supported by the cli and `[ARG...]` is the command name and arguments of the command the user wishes to run on the remote system. Example flow with server address `localhost:3000` and `ping` executable below. 
+### start 
+`start` starts a process on the server. An id identify the process is returned to be used with the other commands. Specify the `-l` flag to listen for process output after starting the process. This will return a job id of `job1` for example.
+```shell
+> teleport --cacert="ca.cert" --key="client.key" --cert="client.cert" --addr="localhost:3000" \
+           start -l ping google.com
+```
+### status
+`status` returns the status of a job. Specify the job id returned from the `start` command to run `status`.
+```shell
+teleport --cacert="ca.cert" --key="client.key" --cert="client.cert" --addr="localhost:3000" status job1
+```
+### output
+`output` displays the output from a process. Specify the job id returned from the `start` command to run `output`.
+```shell
+teleport --cacert="ca.cert" --key="client.key" --cert="client.cert" --addr="localhost:3000" output job1
+```
+### stop
+`stop` stops a process running on the server. Specify the job id returned from the `start` command to run `stop`.
+```shell
+teleport --cacert="ca.cert" --key="client.key" --cert="client.cert" --addr="localhost:3000" stop job1
+```
+The full set of commands will be provided through the help message which can be invoked with:
+```shell
+> teleport --help
+```
+
+## Authentication/Authorization
+Client and server communication must be encrypted and secure. mTLS authentication is used to achieve this. A tool will be provided to generate certificates for the client and server using the same CA. The client certificates are each generated with a unique common name so the server can differentiate between different clients.
```

Comment:
> A tool will be provided to generate certificates for the client and server using the same CA.

It's ok for simplicity to generate certs/keys and include them in the repo if that makes things easier.